### PR TITLE
ENG-17757 use designated container for native direct allocated memory

### PR DESCRIPTION
**Pull request by zeemanhe**
The container is responsible for cleaning up native allocated direct memory.

### CODE REVIEW

File: src/frontend/org/voltcore/utils/DBBPool.java.template [hunk 1]
```diff
@@ -563,7 +583,7 @@ public final class DBBPool {
     private static native void nativeDeleteCharArrayMemory(long pointer);
 
     public static BBContainer allocateUnsafeByteBuffer(long size) {
-        final BBContainer retcont = DBBPool.wrapBB(nativeAllocateUnsafeByteBuffer(size));
+        final BBContainer retcont = new NDBBWrapperContainer(nativeAllocateUnsafeByteBuffer(size));
```
Comment (JadenW) [inline src/frontend/org/voltcore/utils/DBBPool.java.template]: There are other places other than here probably need to be wrap as NDBBWrapperContainer also.
There are 7 other JNI call of NewDirectByteBuffer in JNITopend.cpp. 
For example, https://github.com/VoltDB/voltdb/blob/master/src/frontend/org/voltdb/export/ExportManager.java#L511 was also allocate the native dbb. 
There also the method registerUnsafeMemory, but only for debug.

Comment (zeemanhe) [reply src/frontend/org/voltcore/utils/DBBPool.java.template]: NewDirectByteBuffer on JNITopend would be deallocated within JINTopend.  ExportManager does not seem to have direct memory allocated there.

Comment (n3world) [reply src/frontend/org/voltcore/utils/DBBPool.java.template]: I looked at most of the uses of NewDirectByteBuffer in JNITopend and I think most deallocated the memory but I wasn't sure about the life cycle of fallbackToEEAllocatedBuffer https://github.com/VoltDB/voltdb/blob/master/src/ee/execution/JNITopend.cpp#L289

The only types of buffers that would have been freed previously are ones that were wrapped in a BBContainer so this should fix all of those and any other places with an issue would have been a memory leak in previous versions too.

Comment (zeemanhe) [reply src/frontend/org/voltcore/utils/DBBPool.java.template]: @kaikaizi  and I looked at the buffer from fallbackToEEAllocatedBuffer.  The buffer is set to null  on https://github.com/VoltDB/voltdb/blob/master/src/frontend/org/voltdb/jni/ExecutionEngineJNI.java#L479,  does not look like it is  deallocated. These may be related to large temp table work which could be deprecated according to @wweiss-voltdb .

Comment (n3world) [reply src/frontend/org/voltcore/utils/DBBPool.java.template]: Then a ticket should be filed to remove it so we don't have a potential leak if anybody thinks they could use it in the future.

Comment (zeemanhe) [reply src/frontend/org/voltcore/utils/DBBPool.java.template]: Potential leaks, including fallback EE allocated buffer, will be addressed in ENG-18332

Comment (wweiss-voltdb) [reply src/frontend/org/voltcore/utils/DBBPool.java.template]: result buffers are used to provide the outcome to sql statements. fallbackToEEAllocatedBuffer is used for results that are more than the default result reserved memory area specified in setBuffers(). My recollection is that the reserved space is something like 10MB and the fallbackBuffer is used for more than 10MB results. However, when the transaction finishes, the EE always resets the buffer from the fallbackBuffer back to the reserved buffer and then deallocates the fallbackBuffer. While this path is different from Java managed ee Buffer memory it is still cleaned up.

**Pull request merged**